# An-immunophenotype-coupled_transcriptomic_atlas_of-human_hematopoietic_progenitors
My analysis of some scRNAseq data coupled with Flow Cytometry data of human BM Hematopoietic cell population which was published in Nature Immunology April 2024 (https://doi.org/10.1038/s41590-024-01782-4). Online supplemental content including data and code is also available at https://doi.org/10.1038/s41590-024-01782-4.  
This publicatioin showcases the most comprehensive and updated study characterising cellular components of human bone marrow mononuclear cells (BMMC) in the hematopoietic system as well as non-hematopoietic supporting cells.  To address a overwhelming cellular complexity of the human bone marrow hematopoietic system, authors employed multichromatic flow cytmetric anaysis in conjunction with CITE-seq based transcriptome characterizations of cells from bone marrow aspirates, through which they were able to discern over **40 distinct phenotypic and transcriptomic subsets** within a population of bone marrow mononulear cells.    
My first impression of this publication is that this study was seriously well done.  From the bench experimentation to data analysis, every steps seemingly were accomplished rightly.  The authors even titrated ***275 TotalseqA antibodies against actual BMMCs !!***  Who in the right mind would do this and more practically, who could afford so many BMMCs for titrating 275 antibodies ?  Surely, using BMMCs is perhaps the best and most appropriate way to titrate antibodies but BMMCs are significantly more precious and hard to obtain.  It is certainly not the best but I would substitute PBMCs for BMMCs.  Overall, this whole study is very impressive, and it is the most comprehensive characterization  of human hematopoietic system in the recent years.  
Raw sequencing (fastqs) and processed marix files are available in GEO under accession GSE245108 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE245108). Flow data (though imputed) are available from FlowRepository, FR-FCM-Z6UQ (http://flowrepository.org/id/FR-FCM-Z6UQ).  Also, authors set up a github page where all the main and accessory codes used for the publication  are found (https://github.com/nsalomonis/Human-Bone-Marrow-Titration-Atlas).  In addition, web based analysis and visualization interface is available (https://altanalyze.org/MarrowAtlas/).  
